bindings/cs/MongoDB.Crypt/CryptContext.cs

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2019–present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace MongoDB.Crypt
+{
+    /// <summary>
+    /// A encryption or decryption session. It may not be reused.
+    /// </summary>
+    /// <seealso cref="System.IDisposable" />
+    /// <seealso cref="MongoDB.Crypt.IStatus" />
+    public class CryptContext : IDisposable, IStatus
+    {
+        /// <summary>
+        /// States of the CryptContext state machine
+        /// </summary>
+        public enum StateCode
+        {
+            /// <summary>
+            /// LibMongoCrypt hit an error
+            /// </summary>
+            MONGOCRYPT_CTX_ERROR = 0,
+
+            /// <summary>
+            /// LibMongoCrypt wants the collection information by running a OP_MSG command against the users' mongod.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_MONGO_COLLINFO = 1,
+
+            /// <summary>
+            /// LibMongoCrypt wants a command to be run against mongocryptd.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_MONGO_MARKINGS = 2,
+
+            /// <summary>
+            /// LibMongoCrypt wants a command to be run against mongod key vault.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_MONGO_KEYS = 3,
+
+            /// <summary>
+            /// LibMongoCrypt wants a request sent to KMS.
+            /// </summary>
+            MONGOCRYPT_CTX_NEED_KMS = 4,
+
+            /// <summary>
+            /// LibMongoCrypt is ready to do encryption, call Finish().
+            /// </summary>
+            MONGOCRYPT_CTX_READY = 5,
+
+            /// <summary>
+            /// LibMongoCrypt is complete.
+            /// </summary>
+            MONGOCRYPT_CTX_DONE = 6
+        }
+
+        private ContextSafeHandle _handle;
+        private Status _status;
```

Comment:
Similary unclear about the role of the Status object here also.

Are we simply allocation a status object that we can repeatedly pass to libmongocrypt methods for them to return status information in? 

Is this simply an optimization (instead of allocating a new status object for each call).

Reply:
Yes, it is simply an optimization to avoid allocating and deallocating since malloc/free more expensive then new in the clr